### DISCUSSION TITLE
link error in cross compilation (linux -> macos)

### DISCUSSION THREAD

**Discussion by cstkingkey** [fltk-rs/fltk-rs#1541] [General]
after updating from 1.4.26 to 14.29  
ld: warning: object file (target/x86_64-apple-darwin/debug/deps/libfltk_sys-41e8423397c8e5fb.rlib(Fl_Scroll.cxx.o)) was built for newer macOS version (13.1) than being linked (10.9)

 Undefined symbols for architecture x86_64:
            "_OBJC_CLASS_$_UTType", referenced from:
                objc-class-ref in libfltk_sys-41e8423397c8e5fb.rlib(Fl_Native_File_Chooser_MAC.mm.o)
          ld: symbol(s) not found for architecture x86_64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by MoAlyousef** [comment]
How are you cross-compiling?
FLTK now requires UniformTypeIdentifiers.framework when building on macos > 11.0. The prebuilt binaries are built on macos-latest, but even if we tried to build on macos 10, it’s not available in the github actions runners. 
One option is to prebuild the binaries on macos 10, then pass CFLTK_BUNDLE_URL to your hosted prebuilt binaries.

**Reply by cstkingkey** [reply]
I'm using https://github.com/tpoechtrager/osxcross

**Comment by cstkingkey** [comment]
How can I downgrade it?
I use "fltk = { version = "=1.4.26", features = [] }" to specify fltk version, but it still use fltk-sy-1.4.29

**Reply by MoAlyousef** [reply]
You can add fltk-sys as a dependency and specify it to also be 1.4.26.

However I would like to make things work with osxcross. It seems it uses the macos MACOSX_DEPLOYMENT_TARGET, which the fltk-rs build script could query to determine what macos version is targetted. 
I'll try to push a fix in master in a few minutes, then you could try using
```
[patch.crates-io]
fltk = { git = "https://github.com/fltk-rs/fltk-rs" }
```

**Reply by MoAlyousef** [reply]
I have pushed a fix I think.

**Reply by cstkingkey** [reply]
It's not working. 
I found https://github.com/fltk-rs/fltk-rs/issues/1536, the fltk logic to reference UniformTypeIdentifiers need to be changed.

**Reply by MoAlyousef** [reply]
I changed it to:
```rust
let darwin_version: i32 = utils::get_taget_darwin_major_version().unwrap();
                if darwin_version > 19 {
                    println!("cargo:rustc-link-lib=framework=UniformTypeIdentifiers");
                }
```
Darwin 20 corresponds to macos 11. It gets the darwin version from the MACOSX_DEPLOYMENT_TARGET env variable. From looking at osxcross it seems that it sets the env variable during the build.

**Reply by cstkingkey** [reply]
the target is 10.9, but fltk still require to link UniformTypeIdentifiers

**Reply by MoAlyousef** [reply]
I'm thinking the issue might be upstream. Can you try building FLTK with osxcross:
```
git clone https://github.com/fltk/fltk
cd fltk
cmake -Bbin && cmake --build bin
bin/test/editor &
```

Also if there's some workflow where I can test with osxcross without having to download it or package the sdk. Their readme isn't clear on this.

**Reply by cstkingkey** [reply]
You can use https://github.com/growtopiajaw/MacOSX-SDK/releases/tag/v1.0

**Reply by MoAlyousef** [reply]
I have tried the juanplopes/crossbuild container and it seems FLTK itself fails to build unless you set "ENV MACOSX_DEPLOYMENT_TARGET=11.0.0":
```dockerfile
FROM juanplopes/crossbuild
RUN git clone https://github.com/fltk/fltk --depth=1
ENV CROSS_TRIPLE=x86_64-apple-darwin
WORKDIR /workdir/fltk
ENV MACOSX_DEPLOYMENT_TARGET="10.9.0"
RUN crossbuild cmake -Bbin -DFLTK_BUILD_GL=OFF -DCMAKE_SYSTEM_NAME="Darwin" -DCMAKE_SYSTEM_VERSION="19.0.0"
RUN crossbuild cmake --build bin --parallel
```
Unfortunately crosscompiling to macos is finicky, especially when it comes to sdk’s which are already end-of-life. 
You can try asking in the fltk-general forum: 
https://groups.google.com/g/fltkgeneral

**Reply by cstkingkey** [reply]
https://github.com/fltk/fltk/commit/97d2836f5e72054f640289a93c4f2dd776c4fd00

I think code shall use MAC_OS_X_VERSION_MIN_REQUIRED

> MAC_OS_X_VERSION_MIN_REQUIRED specifies the earliest OS version you support, and corresponds to MACOSX_DEPLOYMENT_TARGET in your project/ target configuration.

> MAC_OS_X_VERSION_MAX_ALLOWED specifies the maximum OS version your code knows about, and corresponds to SDKROOT in your project/target configuration.

**Reply by MoAlyousef** [reply]
I tried replacing those but it doesn't change the result:
```dockerfile
FROM juanplopes/crossbuild
RUN git clone https://github.com/fltk/fltk --depth=1
ENV CROSS_TRIPLE=x86_64-apple-darwin
WORKDIR /workdir/fltk
ENV MACOSX_DEPLOYMENT_TARGET="10.9.0"
RUN sed -i '31s/MAC_OS_X_VERSION_MAX_ALLOWED/MAC_OS_X_VERSION_MIN_REQUIRED/' src/Fl_Native_File_Chooser_MAC.mm
RUN head -34 src/Fl_Native_File_Chooser_MAC.mm
RUN crossbuild cmake -Bbin -DFLTK_BUILD_GL=OFF -DCMAKE_SYSTEM_NAME="Darwin" -DCMAKE_SYSTEM_VERSION="19.0.0"
RUN crossbuild cmake --build bin --parallel
```

I'll try asking at the fltk-general forum.

**Reply by cstkingkey** [reply]
according to the comment in https://github.com/fltk/fltk/pull/965
the UniformTypeIdentifiers shall be linked with newer sdk. it's not related to MACOSX_DEPLOYMENT_TARGET.

**Reply by MoAlyousef** [reply]
MACOSX_DEPLOYMENT_TARGET sets CMAKE_OSX_DEPLOYMENT_TARGET according to the cmake docs. On the fltk-rs side, the thing I can offer is to set the CMAKE_SYSTEM_VERSION to 20 when cross-compiling from linux to macos when MACOSX_DEPLOYMENT_TARGET is not set or to just ignore MACOSX_DEPLOYMENT_TARGET even if it's set.

**Reply by MoAlyousef** [reply]
The following container will build when MACOSX_DEPLOYMENT_TARGET is set to 11 for example:
```dockerfile
FROM juanplopes/crossbuild
RUN git clone https://github.com/fltk-rs/fltk-rs --depth=1 --recurse-submodules
ENV CROSS_TRIPLE=x86_64-apple-darwin
WORKDIR /workdir/fltk-rs
ENV RUSTUP_HOME="/usr/local/rustup" CARGO_HOME="/usr/local/cargo" PATH="/usr/local/rustup/bin:/usr/local/cargo/bin:$PATH"
RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain stable --profile minimal -y
RUN rustup target add x86_64-apple-darwin
RUN rm -rf /usr/bin/ld
RUN ln -s /usr/x86_64-apple-darwin20.4/bin/ld /usr/bin/ld
ENV MACOSX_DEPLOYMENT_TARGET="11" CC=/usr/x86_64-apple-darwin20.4/bin/cc CXX=/usr/x86_64-apple-darwin20.4/bin/c++ CARGO_TARGET_X86_64_APPLE_DARWIN_LINKER=/usr/x86_64-apple-darwin20.4/bin/cc
RUN cargo build --example hello_button --target=x86_64-apple-darwin
```
The issue is that the toolchain is darwin20.4, which has the UniformTypeIdentifier header so it will compile it, and if a lower deployment target is set, for example 10.9, it will cause a link error. 
The issue primarily lies with how macos sdk's work I believe.

**Comment by cstkingkey** [comment]
> The following container will build when MACOSX_DEPLOYMENT_TARGET is set to 11 for example:
> 
> ```dockerfile
> FROM juanplopes/crossbuild
> RUN git clone https://github.com/fltk-rs/fltk-rs --depth=1 --recurse-submodules
> ENV CROSS_TRIPLE=x86_64-apple-darwin
> WORKDIR /workdir/fltk-rs
> ENV RUSTUP_HOME="/usr/local/rustup" CARGO_HOME="/usr/local/cargo" PATH="/usr/local/rustup/bin:/usr/local/cargo/bin:$PATH"
> RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain stable --profile minimal -y
> RUN rustup target add x86_64-apple-darwin
> RUN rm -rf /usr/bin/ld
> RUN ln -s /usr/x86_64-apple-darwin20.4/bin/ld /usr/bin/ld
> ENV MACOSX_DEPLOYMENT_TARGET="11" CC=/usr/x86_64-apple-darwin20.4/bin/cc CXX=/usr/x86_64-apple-darwin20.4/bin/c++ CARGO_TARGET_X86_64_APPLE_DARWIN_LINKER=/usr/x86_64-apple-darwin20.4/bin/cc
> RUN cargo build --example hello_button --target=x86_64-apple-darwin
> ```
> 
> The issue is that the toolchain is darwin20.4, which has the UniformTypeIdentifier header so it will compile it, and if a lower deployment target is set, for example 10.9, it will cause a link error. The issue primarily lies with how macos sdk's work I believe.

Is there is a way to detect sdk version, for example a env var? If I link it unconditionally, the link error is gone. but currently I don't have the machine to test whether it can run on macos 10.13.

**Reply by MoAlyousef** [reply]
The `fl_mac_os_version >= 110000` runtime check should allow it to run on older versions. 

I’m not sure there is an easy way to detect the sdk version with osxcross. 
On macos you could always use `xcrun --show-sdk-version`.

**Comment by cstkingkey** [comment]
https://github.com/fltk/fltk/commit/70b84b491b889ecb528e46aae20b6105e0b96065
fltk has a fix

**Reply by MoAlyousef** [reply]
Thanks for the heads up. I've pulled the fix from fltk/fltk. Can you try building with master.

**Reply by MoAlyousef** [reply]
Assuming from [this comment](https://github.com/fltk/fltk/pull/965#issuecomment-2094152137) that this now works.

**Reply by cstkingkey** [reply]
Thanks for your help. When will the new version be published?

**Reply by MoAlyousef** [reply]
You’re welcome. In 2 to 3 days.
